chore: replace occurrences of 'localhost' with '127.0.01'; use random ports; closes #15399

boneskull opened the pull request:
Because `localhost` can resolve to an IPv6 address and there is some sort of incompatibility (@mykola-mokhnach?) we want to ensure we're binding to an IPv4 address.

This also adds more support for random ports, where applicable.  `fake-driver` is _not yet_ applicable.

boneskull committed:
chore: replace occurrences of 'localhost' with '127.0.01'; use random ports; closes #15399

- `base-driver` gets a new `helpers` module with a `getTestPort` function; this is now used throughout the tests
- `fake-driver` can't use random ports because it depends on `appium-base-driver`'s test suite, and that is _not_ the current appium-2.0-compatible `@appium/base-driver`

jlipps reviewed (approved):
saw some github actions check warnings for the first time but assuming lint passes seems good to go!

mykola-mokhnach commented on packages/base-driver/test/basedriver/driver-e2e-tests.js:
```diff
@@ -1,57 +1,60 @@
 function baseDriverE2ETests (DriverClass, defaultCaps = {}) {
+  const DEFAULT_ARGS = {
+    address: TEST_HOST,
+    port: null
```
I would add a short comment about the real meaning of null value here

boneskull replied:
this is outdated

mykola-mokhnach commented on packages/base-driver/test/basedriver/driver-e2e-tests.js:
```diff
@@ -464,7 +467,7 @@ function baseDriverE2ETests (DriverClass, defaultCaps = {}) {
         `;
         const {data} = await axios({
-          url: `http://localhost:${port}/session/${sessionId}/appium/execute_driver`,
+          url: `http://${TEST_HOST}:${port}/session/${sessionId}/appium/execute_driver`,
```
I would rather create a small function to make such urls, where we pass port number, path and session Id. It could even be curried for particular suites, so less arguments are needed to be passed

mykola-mokhnach commented on packages/base-driver/test/jsonwp-proxy/url-specs.js:
```diff
@@ -3,108 +3,123 @@
+
+  describe('proxying full urls', function () {
+    it('should translate host and port', function () {
+      let incomingUrl = 'http://127.0.0.2:4723/status';
```
should this url be moved to a constant?

jlipps commented:
@boneskull did you want to rebase this?

boneskull committed:
chore(fake-driver): make wdio quieter

boneskull committed:
fix(base-driver): better URL handling in driver-e2e tests

- fix a potential port-reuse issue in `getTestPort()` helper
- dedupe `127.0.0.1` constant
- export curried `createAppiumURL(address, port, session, pathname)` helper for less error-prone URL generation
- export some HTTP method names as constants

boneskull committed:
fix(base-driver): allow https in helper URL generation

- use constants/helpers for `jsonwp-proxy` url tests

boneskull committed:
chore(base-driver): revert name change of TEST_HOST to DEFAULT_TEST_HOST

This would have been a breaking change if other packages depend on it, since I think it's a public API

boneskull commented:
this is ready

boneskull commented:
note that not _all_ tests which _could_ use the curried URL-builder actually do so (in this PR)

jlipps reviewed (approved):
@boneskull want to rebase this so we can merge?

jlipps commented:
@boneskull another ping on this, want to rebase and fix conflicts so we can merge?

boneskull commented:
will do

boneskull committed:
chore(pkg): obviousify scripts
- also speedify install by running `postinstall` builds in parallel

boneskull commented:
gotta sign the corporate CLA... waiting for legal to give me green light

boneskull commented:
@jlipps OK, got the CLA sorted.  merge?
